pkg/mimirtool/commands/remote_read.go

```diff
@@ -247,36 +257,353 @@ func (c *RemoteReadCommand) prepare() (query func(context.Context) (storage.Seri
+		}
+
+		pbQuery, err := remote.ToQuery(
+			int64(model.TimeFromUnixNano(from.UnixNano())),
+			int64(model.TimeFromUnixNano(to.UnixNano())),
+			matchers,
+			nil,
+		)
+		if err != nil {
+			return nil, time.Time{}, time.Time{}, fmt.Errorf("error creating query for selector '%s': %w", selector, err)
+		}
+		pbQueries = append(pbQueries, pbQuery)
 	}
 
 	readClient, err := c.readClient()
 	if err != nil {
 		return nil, time.Time{}, time.Time{}, err
 	}
 
-	pbQuery, err := remote.ToQuery(
-		int64(model.TimeFromUnixNano(from.UnixNano())),
-		int64(model.TimeFromUnixNano(to.UnixNano())),
-		matchers,
-		nil,
-	)
+	return func(ctx context.Context) (storage.SeriesSet, error) {
+		log.Infof("Querying time from=%s to=%s with %d selectors", from.Format(time.RFC3339), to.Format(time.RFC3339), len(c.selectors))
+		for i, selector := range c.selectors {
+			log.Debugf("Selector %d: %s", i+1, selector)
+		}
+		return c.executeMultipleQueries(ctx, readClient, pbQueries)
+	}, from, to, nil
+}
+
+// executeMultipleQueries sends multiple queries in a single protobuf request
+func (c *RemoteReadCommand) executeMultipleQueries(ctx context.Context, readClient remote.ReadClient, queries []*prompb.Query) (storage.SeriesSet, error) {
+	// We'll use reflection to access the private fields of the client to send a batched request
+	// This is hacky but gets the job done for now
+	client, ok := readClient.(*remote.Client)
+	if !ok {
+		return nil, fmt.Errorf("unexpected readClient type: %T", readClient)
+	}
+
+	// Build the batched request with user-selected response type preference
+	var acceptedTypes []prompb.ReadRequest_ResponseType
+	if c.useChunks {
+		log.Debugf("Requesting chunked streaming response (STREAMED_XOR_CHUNKS)")
+		acceptedTypes = []prompb.ReadRequest_ResponseType{
+			prompb.ReadRequest_STREAMED_XOR_CHUNKS,
+			prompb.ReadRequest_SAMPLES, // fallback
+		}
+	} else {
+		log.Debugf("Requesting sampled response (SAMPLES)")
+		acceptedTypes = []prompb.ReadRequest_ResponseType{
+			prompb.ReadRequest_SAMPLES,
+		}
+	}
+
+	req := &prompb.ReadRequest{
+		Queries:               queries,
+		AcceptedResponseTypes: acceptedTypes,
+	}
+
+	// Marshal the batched request
+	data, err := proto.Marshal(req)
 	if err != nil {
-		return nil, time.Time{}, time.Time{}, err
+		return nil, fmt.Errorf("unable to marshal read request: %w", err)
 	}
 
-	return func(ctx context.Context) (storage.SeriesSet, error) {
-		log.Infof("Querying time from=%s to=%s with selector=%s", from.Format(time.RFC3339), to.Format(time.RFC3339), c.selector)
-		resp, err := readClient.Read(ctx, pbQuery, false)
+	compressed := snappy.Encode(nil, data)
+
+	// Use reflection to get the URL from the client
+	clientValue := reflect.ValueOf(client).Elem()
+	urlStringField := clientValue.FieldByName("urlString")
+	if !urlStringField.IsValid() {
+		return nil, fmt.Errorf("unable to access urlString field")
+	}
+	urlString := urlStringField.String()
```

Comment:
Why do we need to use reflection here? Can't we use the `address` and `remoteReadPath` fields on `RemoteReadCommand`?